text/0000-time-improvements.md

```diff
@@ -0,0 +1,330 @@
+dates and times. It is intentionally forwards-compatible with such
+extensions.
+
+---
+
+Finally, many APIs that **take** a `Duration` can only do something useful with
+positive values. For example, a timeout API would not know how to wait a
+negative amount of time before timing out. Even discounting the possibility of
+coding mistakes, the problem of system clock time travel means that programmers
+often produce negative durations that they did not expect, and APIs that
+liberally accept negative durations only propagate the error further.
+
+As a result, this RFC makes a number of simplifying assumptions that can be
+relaxed over time with additional types or through further RFCs:
+
+It provides convenience methods for constructing Durations from larger units
+of time (minutes, hours, days), but gives them names like
+`Duration.from_standard_hour`. A standard hour is always 3600 seconds,
+regardless of leap seconds.
+
+It provides APIs that are expected to produce positive `Duration`s, and expects
+that APIs like timeouts will accept positive `Durations` (which is currently
+the case in Rust's standard library). These APIs help the programmer discover
+the possibility of system clock time travel, and either handle the error explicitly,
+or at least avoid propagating the problem into other APIs (by using `unwrap`).
+
+It separates monotonic time (`ProcessTime`) from time derived from the system
+clock (`SystemTime`), which must account for the possibility of time travel.
+This allows methods related to monotonic time to be uncaveated, while working
+with the system clock has more methods that return `Result`s.
+
+This RFC does not attempt to define a type for calendared DateTimes, nor does it
+directly address time zones.
+
+# Proposal
+
+## Types
+
+```rs
+pub struct ProcessTime {
+  secs: u64,
+  nanos: u32
+}
+
+pub struct SystemTime {
+  secs: u64,
+  nanos: u32
+}
+
+pub struct Duration {
+  secs: u64,
+  nanos: u32
+}
+```
+
+### ProcessTime
+
+`ProcessTime` is the simplest of the instant types. It represents an opaque
+(non-serializable!) timestamp that is guaranteed to be monotonic throughout
+the timeframe of the process it was created in.
+
+> In this context, monotonic means that a timestamp created later in real-world
+> time will always be larger than a timestamp created earlier in real-world
+> time.
+
+The `Duration` type can be used in conjunction with `ProcessTime`, and these
+operations have none of the usual time-related caveats.
+
+* Add a `Duration` to a `ProcessTime`, producing a new `ProcessTime`
+* compare two `ProcessTime`s to each other
+* subtract a `ProcessTime` from a later `ProcessTime`, producing a `Duration`
+* ask for an amount of time elapsed since a `ProcessTime`, producing a `Duration`
+
+Asking for an amount of time elapsed from a given `ProcessTime` is a very common
+operation that is guaranteed to produce a positive `Duration`. Asking for the
```

Comment:
Should this really be `positive`? Or just `non-negative`? (ie: might help to clarify if a elapsed time can be 0)